Source/PluginProcessor.h

```diff
@@ -15,6 +15,163 @@
 #include "Tract.hpp"
 #include "WhiteNoise.hpp"
 #include "Biquad.hpp"
+#include <math.h>
+
+
+class PinkTromboneADSR
+{
+public:
+	//==============================================================================
+	PinkTromboneADSR()
+	{
+		recalculateRates();
+	}
+
+	//==============================================================================
+	struct Parameters
+	{
+		Parameters() = default;
+
+		Parameters (float attackTimeSeconds,
+					float decayTimeSeconds,
+					float sustainLevel,
+					float decayExponent)
+			: attack (attackTimeSeconds),
+			  decay (decayTimeSeconds),
+			  sustain (sustainLevel),
+			  decayExp(decayExponent)
+		{
+		}
+
+		float attack = 0.1f, decay = 0.1f, sustain = 1.0f, attackExp=1.0f, decayExp=1.0f;
+	};
+
+	void setParameters (const Parameters& newParameters)
+	{
+		// need to call setSampleRate() first!
+		jassert (sampleRate > 0.0);
+
+		parameters = newParameters;
+		recalculateRates();
+	}
+
+	//==============================================================================
+	void setSampleRate (double newSampleRate) noexcept
+	{
+		jassert (newSampleRate > 0.0);
+		sampleRate = newSampleRate;
+	}
+
+	//==============================================================================
+
+	void reset() noexcept
+	{
+		envelopeVal = 0.0f;
+		state = State::idle;
+	}
+	
+	void noteOn() noexcept
+	{
+		if (attackRate > 0.0f)
+		{
+			state = State::attack;
+		}
+		else if (decayRate > 0.0f)
+		{
+			envelopeVal = 1.0f;
+			state = State::decay;
+		}
+		else
+		{
+			envelopeVal = parameters.sustain;
+			state = State::sustain;
+		}
+	}
+	
+	void noteOff() noexcept
+	{
+		if (state != State::idle)
+			reset();
+	}
+
+	//==============================================================================
+	float getNextSample() noexcept
+	{
+		float scaledEnvelopeVal;
+		
+		if (state == State::idle)
+			return 0.0f;
+
+		if (state == State::attack)
+		{
+			envelopeVal += attackRate;
+
+			if (envelopeVal >= 1.0f)
+			{
+				envelopeVal = 1.0f;
+				goToNextState();
+			}
+		}
+		else if (state == State::decay)
+		{
+			envelopeVal -= decayRate;
+
+			if (envelopeVal <= parameters.sustain)
+			{
+				envelopeVal = parameters.sustain;
+				goToNextState();
+			}
+			if (parameters.decayExp >= 0)
+				scaledEnvelopeVal = powf(envelopeVal, parameters.decayExp);

```

Comment:
Have you checked the math on this? I think it's not quite right.

Say for example the sustain height is 0.5 and the decayExp is 2. That means that the envelope (during the decay phase) should decay from 1 to 0.5, but along a convex curve. However, applying `powf` when the envelope is near 0.5 will produce an output for `scaledEnvelopeVal` that is close to 0.25. Then, once the envelope decays to 0.5, the envelope will jump back to 0.5.

I think the correct way to do this is to "pretend" that the sustain height is always zero, so that envelopeVal—before being passed to powf—will always be between 1 and 0. Then, after applying the curve and getting scaledEnvelopeVal, scale the output using (1 - sustain) and then bias it by (sustain).